File: ios/Detonator/Sources/Detonator/Module/UtilityModule.swift
import Foundation
import UIKit
import Photos

class UtilityModule: Module {
    public override func setUp() -> Void {
        detonator.setEventListener("com.iconshot.detonator::log") { value in
            let str = String(value.dropFirst().dropLast())
            
            print(str)
        }
        
        detonator.setRequestListener("com.iconshot.detonator::openUrl") { promise, value, edge in
            guard let url = URL(string: value) else {
                promise.reject("Not a valid url.")
                
                return
            }
            
            if !UIApplication.shared.canOpenURL(url) {
                promise.reject("Can't open url.")
                
                return
            }
            
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
            
            promise.resolve()
        }
        
        detonator.setRequestListener("com.iconshot.detonator.imagesize::getSize") { promise, value, edge in
            let source = value
            
            if source.starts(with: "file://") {
                guard let fileURL = URL(string: source) else {
                    promise.reject("Not a valid source.")

                    return
                }
                
                guard let data = try? Data(contentsOf: fileURL) else {
                    promise.reject("Could not load file from local path.")
                    
                    return
                }
                
                guard let image = UIImage(data: data) else {
                    promise.reject("File is not an image.")
                    
                    return
                }
                
                let size = image.size
                
                let imageSize = ImageSize(width: Int(size.width), height: Int(size.height))
                
                promise.resolve(imageSize)
                
                return
            }
            
            if source.starts(with: "content://") {
                let localIdentifier = source.replacingOccurrences(of: "content://", with: "")
                
                let assets = PHAsset.fetchAssets(withLocalIdentifiers: [localIdentifier], options: nil)
                    
                guard let asset = assets.firstObject else {
                    promise.reject("Asset not found.")
                    
                    return
                }
                
                if asset.mediaType != .image {
                    promise.reject("Asset is not an image.")
                    
                    return
                }
                
                let size = ImageSize(width: asset.pixelWidth, height: asset.pixelHeight)
                
                promise.resolve(size)
                
                return
            }
            
            ImageHelper.getImageSize(urlString: source) { size, error in
                if let error = error {
                    promise.reject(error)
                    
                    return
                }
                
                let imageSize = ImageSize(width: Int(size!.width), height: Int(size!.height))
                
                promise.resolve(imageSize)
            }
        }
    }
    
    private struct ImageSize: Encodable {
        let width: Int
        let height: Int
    }
}

File: android/Detonator/src/main/java/com/iconshot/detonator/module/UtilityModule.java
package com.iconshot.detonator.module;

import android.content.ActivityNotFoundException;
import android.content.ContentResolver;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.net.Uri;
import android.os.AsyncTask;
import android.widget.Toast;

import com.iconshot.detonator.Detonator;
import com.iconshot.detonator.helpers.ImageHelper;

import java.io.File;
import java.io.InputStream;

public class UtilityModule extends Module {
    public UtilityModule(Detonator detonator) {
        super(detonator);
    }

    @Override
    public void setUp() {
        detonator.setEventListener("com.iconshot.detonator::log", value -> {
            String str = value.substring(1, value.length() - 1);

            System.out.println(str);
        });

        detonator.setRequestListener("com.iconshot.detonator::openUrl", (promise, value, edge) -> {
            try {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(value));

                detonator.context.startActivity(browserIntent);

                promise.resolve();
            } catch (ActivityNotFoundException e) {
                promise.reject("Can't open url.");
            }
        });

        detonator.setRequestListener("com.iconshot.detonator.imagesize::getSize", (promise, value, edge) -> {
            String source = value;

            if (source.startsWith("file://")) {
                String path = source.substring(7);

                File file = new File(path);

                if (!file.exists()) {
                    promise.reject("File does not exist.");

                    return;
                }

                ImageHelper.Size size = ImageHelper.getSize(file);

                promise.resolve(size);

                return;
            }

            if (source.startsWith("content://")) {
                Uri uri = Uri.parse(source);

                ContentResolver contentResolver = detonator.context.getContentResolver();

                BitmapFactory.Options options = new BitmapFactory.Options();

                options.inJustDecodeBounds = true;

                try (InputStream input = contentResolver.openInputStream(uri)) {
                    BitmapFactory.decodeStream(input, null, options);

                    int width = options.outWidth;
                    int height = options.outHeight;

                    ImageHelper.Size size = new ImageHelper.Size();

                    size.width = width;
                    size.height = height;

                    promise.resolve(size);
                } catch (Exception e) {
                    promise.reject("Unable to get size.");
                }

                return;
            }

            Bitmap cachedBitmap = ImageHelper.getCachedBitmap(detonator.context, source);

            if (cachedBitmap != null) {
                File file = ImageHelper.getFile(detonator.context, source);

                System.out.println(file);

                ImageHelper.Size size = ImageHelper.getSize(file);

                promise.resolve(size);

                return;
            }

            class DownloadImageTask extends AsyncTask<String, Void, File> {
                @Override
                protected File doInBackground(String... urls) {
                    String imageUrl = urls[0];

                    Bitmap bitmap = ImageHelper.loadImageBitmap(detonator.context, imageUrl);

                    if (bitmap == null) {
                        return null;
                    }

                    File file = ImageHelper.getFile(detonator.context, imageUrl);

                    return file;
                }

                @Override
                protected void onPostExecute(File file) {
                    if (file == null) {
                        promise.reject("Couldn't get image size.");

                        return;
                    }

                    ImageHelper.Size size = ImageHelper.getSize(file);

                    promise.resolve(size);
                }
            }

            new DownloadImageTask().execute(source);
        });

        detonator.setRequestListener("com.iconshot.detonator.toast::show", (promise, value, edge) -> {
            ShowToastData data = detonator.decode(value, ShowToastData.class);

            int duration = data.isShort ? Toast.LENGTH_SHORT : Toast.LENGTH_LONG;

            Toast.makeText(detonator.context, data.text, duration).show();

            promise.resolve();
        });
    }

    private static class ShowToastData {
        String text;
        boolean isShort;
    }
}
